GarNet/garnet.py

```diff
@@ -226,23 +153,34 @@ def TF_regression(motifs_and_genes_file_or_dataframe, expression_file, output_di
 		# Occasionally there's only one gene associated with a TF, which we can't fit a line to.
 		if len(expression_profile) < 5: continue
 
+		# This allows heavier points to be visualized on the top instead of being hidden by long distance points 
+		# expression_profile = expression_profile.reindex(expression_profile.motif_gene_distance.abs().sort_values(inplace=False, ascending=False).index)
+
 		# Ordinary Least Squares linear regression
 		result = linear_regression(formula="expression ~ motifScore", data=expression_profile).fit()
 
 		if output_dir:
 			plot = plot_regression(model_results=result, ax=expression_profile.plot(x="motifScore", y="expression", kind="scatter", grid=True))
+			# Add color to points based on distance to gene
+			plt.scatter(expression_profile["motifScore"], expression_profile["expression"],
+				c=[abs(v) for v in expression_profile["motif_gene_distance"].tolist()],
+				norm=matplotlib.colors.LogNorm(vmin=1, vmax=100000, clip=True), 
+				cmap=matplotlib.cm.Blues_r)
+			plt.title("%s, %0.4f" %(TF_name, result.pvalues['motifScore']))
+			plt.colorbar()
+
 			os.makedirs(os.path.join(output_dir, "regression_plots"), exist_ok=True)
-			plot.savefig(os.path.join(output_dir, "regression_plots", TF_name + '.jpg'))
+			plot.savefig(os.path.join(output_dir, "regression_plots", TF_name.replace("/", "-") + '.pdf'))
 
-		imputed_TF_features.append((TF_name, result.params['motifScore'], result.pvalues['motifScore'], expression_profile['geneName'].tolist()))
+		imputed_TF_features.append((TF_name, result.params['motifScore'], result.pvalues['motifScore'], ','.join(expression_profile['geneName'].tolist())))
 
 	imputed_TF_features_dataframe = pd.DataFrame(imputed_TF_features, columns=["Transcription Factor", "Slope", "P-Value", "Targets"])
 
 	# If we're supplied with an output_dir, we'll put a summary html file in there as well.
-	if output_dir:
-		html_output = templateEnv.get_template("summary.jinja").render(images_dir=os.path.join(output_dir,"regression_plots",""), TFs=sorted(imputed_TF_features, key=lambda x: x[2]))
-		with open(os.path.join(output_dir,"summary.html"), "w") as summary_output_file:
-			summary_output_file.write(html_output)
+	# if output_dir:
+	# 	html_output = templateEnv.get_template("summary.jinja").render(images_dir=os.path.join(output_dir,"regression_plots",""), TFs=sorted(imputed_TF_features, key=lambda x: x[2]))
+	# 	with open(os.path.join(output_dir,"summary.html"), "w") as summary_output_file:
+	# 		summary_output_file.write(html_output)
```

Comment:
Why is this commented out? I think we still want to build an html page if an output dir is supplied...

Reply:
I can't get it to work.

Reply:
Okay, I think it makes sense to resolve that before merging this. What do you think? 

What part isn't working/what error message are you getting?

Reply:
I'm getting this error: 

```
>>> results = garnet.TF_regression(df, expression_files, ".")
02:50:24 - GarNet: INFO - Performing linear regression for 7 transcription factor expression profiles...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nfs/latdata/iamjli/packages/GarNet2/GarNet/garnet.py", line 183, in TF_regression
    html_output = templateEnv.get_template("summary.jinja").render(images_dir=os.path.join(output_dir,"regression_plots",""), TFs=sorted(imputed_TF_features, key=lambda x: x[2]))
  File "/nfs/latdata/iamjli/packages/GarNet2/venv/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/nfs/latdata/iamjli/packages/GarNet2/venv/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/nfs/latdata/iamjli/packages/GarNet2/venv/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/nfs/latdata/iamjli/packages/GarNet2/venv/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "./summary.jinja", line 21, in top-level template code
    {% for Transcription_Factor, Slope, PVal in TFs %}
ValueError: too many values to unpack (expected 3)
```
The dataframe has 4 columns: Transcription Factor, Slope, P-Value , Targets. 

I get the same error when the dataframe uses list objects and string in the Targets column.

Reply:
Ahh, so this is because of the addition of Targets. Would you mind trying changing
```
    {% for Transcription_Factor, Slope, PVal in TFs %}
```
to
```
    {% for Transcription_Factor, Slope, PVal, Targets in TFs %}
```
If that doesn't work just comment it out and I'll fix it.

Reply:
Okay thanks, the html file is generated now, but the plots do not appear.

Reply:
Ah, was printing figs as PDFs, changed back to png.